### PR TITLE
fix(cxx_common): attempt to reduce memory usage of bazel artifact selector

### PR DESCRIPTION
**Pull request by shahms**
This should reduce memory usage somewhat, although there are likely more gains to be had particularly during serialization.  This should incidentally reduce the size of the serialized state by removing the unused fields and using a slightly more compact format for the fields which *are* used.

### CODE REVIEW

File: kythe/cxx/extractor/bazel_artifact_selector.h [hunk 1]
```diff
@@ -168,13 +174,19 @@ class AspectArtifactSelector final : public BazelArtifactSelector {
   absl::Status DeserializeFrom(const google::protobuf::Any& state) final;
 
  private:
+  struct FileSet {
+    std::vector<const BazelArtifactFile*> files;
```
Comment (jaysachs) [inline kythe/cxx/extractor/bazel_artifact_selector.h:178]: The `files` hash_map owns these objects, right?

File: kythe/cxx/extractor/bazel_artifact_selector.cc [hunk 2]
```diff
@@ -214,19 +238,22 @@ void AspectArtifactSelector::ReadFilesInto(
...
-    for (const auto& file : filesets.files()) {
-      files.push_back({
-          .local_path = AsLocalPath(file),
-          .uri = AsUri(file),
-      });
+    const FileSet& fileset = node.mapped();
+    files.reserve(files.size() + fileset.files.size());
+    for (const auto* file : fileset.files) {
+      auto iter = state_.files.find(*file);
+      CHECK(iter != state_.files.end()) << "Attempt to remove a missing file!";
```
Comment (jaysachs) [inline kythe/cxx/extractor/bazel_artifact_selector.cc:247]: So if we get bad data we crash? i.e. this is an unrecoverable state?

File: kythe/cxx/extractor/bazel_artifact_selector.cc [hunk 3]
```diff
@@ -214,19 +238,22 @@ void AspectArtifactSelector::ReadFilesInto(
...
     state_.disposed.insert(std::string(id));
-    auto node = state_.filesets.extract(iter);
-    const build_event_stream::NamedSetOfFiles& filesets = node.mapped();
-
-    for (const auto& file : filesets.files()) {
-      files.push_back({
-          .local_path = AsLocalPath(file),
-          .uri = AsUri(file),
-      });
+    const FileSet& fileset = node.mapped();
```
Comment (jaysachs) [inline kythe/cxx/extractor/bazel_artifact_selector.cc:243]: I'll just note that with everything named "file" or "iter" it's a little challenging to follow what's going on.

**Review by jaysachs — COMMENTED** [review]
Are there any tests that could be updated or written?

File: kythe/cxx/extractor/bazel_artifact_selector.h (hunk 1 above)
Comment (shahms) [reply kythe/cxx/extractor/bazel_artifact_selector.h:178]: Yeah, that's right. I'll add a comment to that effect in a separate PR.

File: kythe/cxx/extractor/bazel_artifact_selector.cc (hunk 2 above)
Comment (shahms) [reply kythe/cxx/extractor/bazel_artifact_selector.cc:247]: This is unrecoverable. The `files` set owns the data. If we fail to lookup the record from a pointer to that record, we're well in UB territory and thoroughly screwed.

**Comment by shahms** [comment]
> Are there any tests that could be updated or written?

The user-visible behavior here is intentionally identical so there's not really anything to test at the moment.

**Pull request merged**